Mathlib/Lean/Expr/ReplaceRec.lean

```diff
@@ -30,6 +30,22 @@ def replaceRec (f? : (Expr → Expr) → Expr → Option Expr) : Expr → Expr :
     | some x => x
     | none   => traverseChildren (M := Id) r e
 
+/-- replaceRec under a monad. -/
+def replaceRecM [Monad M] (f? : (Expr → M Expr) → Expr → M (Option Expr)) : Expr → M Expr :=
+  fix fun r e => do
+    match ← f? r e with
+    | some x => return x
+    | none => traverseChildren r e
+
+/-- replaceRec except that bound variables are instantiated with free variables.
+This means that MetaM tactics can be used inside the replacement function. -/
+def replaceRecMeta [Monad M] [MonadLiftT MetaM M] [MonadControlT MetaM M]
```

Comment:
You might want to look at `Lean.Meta.transform`.

Reply:
Yeah Lean.Meta.Transform doesn't let you recursively call the transform function which led to issues when porting to_additive